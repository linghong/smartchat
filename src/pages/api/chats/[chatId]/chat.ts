import 'reflect-metadata';
import { NextApiRequest, NextApiResponse } from 'next';
import {
  DataSource,
  EntityNotFoundError,
  ConnectionIsNotSetError,
  EntityMetadataNotFoundError
} from 'typeorm';

import { getAppDataSource, Chat, User, ChatMessage, ChatImage } from '@/src/db';

const handleDeleteRequest = async (
  res: NextApiResponse,
  dataSource: DataSource,
  chatIdNum: number
) => {
  // Start a transaction
  await dataSource.transaction(async transactionalEntityManager => {
    // Find the chat
    const chat = await transactionalEntityManager.findOne(Chat, {
      where: { id: chatIdNum },
      relations: ['messages', 'messages.images']
    });

    if (!chat) {
      return res.status(404).json({ message: 'Chat not found' });
    }

    // Delete all images associated with the chat's messages
    if (chat.messages) {
      for (const message of chat.messages) {
        if (message.images) {
          await transactionalEntityManager.delete(ChatImage, {
            messageId: message.id
          });
        }
      }
    }

    // Delete all messages associated with the chat
    await transactionalEntityManager.delete(ChatMessage, { chatId: chat.id });

    // Finally, delete the chat itself
    await transactionalEntityManager.delete(Chat, { id: chat.id });
  });

  res
    .status(200)
    .json({ message: 'Chat and all associated data deleted successfully' });
};

const handlePutRequest = async (
  req: NextApiRequest,
  res: NextApiResponse,
  dataSource: DataSource,
  chatIdNum: number
) => {
  const { title } = req.body;
  const chatRepository = dataSource.getRepository(Chat);

  if (!title) {
    return res.status(400).json({ message: 'Title is required' });
  }

  if (typeof title !== 'string' || title.length < 1 || title.length > 255) {
    return res
      .status(400)
      .json({ message: 'Title must be a string between 1 and 255 characters' });
  }

  // First, find the chat
  const chat = await chatRepository.findOne({ where: { id: chatIdNum } });

  if (!chat) {
    return res.status(404).json({ message: 'Chat not found' });
  }

  // Update the chat title
  chat.title = title;
  await chatRepository.save(chat);

  res.status(200).json({ message: 'Chat title updated successfully' });
};

const handler = async (req: NextApiRequest, res: NextApiResponse) => {
  // validation
  const { chatId } = req.query;
  const chatIdNum = Number(chatId);
  if (isNaN(chatIdNum)) {
    return res.status(400).json({ message: 'Invalid chat ID' });
  }

  try {
    const dataSource = await getAppDataSource();

    if (!dataSource)
      return res.status(500).json({ message: 'AppDataSource is null' });

    switch (req.method) {
      case 'DELETE':
        await handleDeleteRequest(res, dataSource, chatIdNum);
        break;
      case 'PUT':
        await handlePutRequest(req, res, dataSource, chatIdNum);
        break;
      default:
        res.setHeader('Allow', ['DELETE', 'PUT']);
        res.status(405).end(`Method ${req.method} Not Allowed`);
    }
  } catch (error) {
    console.error('Error during Chat operation', error);
    if (error instanceof EntityNotFoundError) {
      res.status(404).json({ message: 'Chat not found' });
    } else if (error instanceof ConnectionIsNotSetError) {
      res.status(500).json({ message: 'Database connection error' });
    } else if (error instanceof EntityMetadataNotFoundError) {
      res.status(500).json({ message: 'Entity metadata error' });
    } else {
      res.status(500).json({ message: 'Internal server error' });
    }
  }
};

export default handler;
