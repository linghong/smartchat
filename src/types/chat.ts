export interface Chat {
  id: number;
  title: string;
}

export type ChatRole = 'system' | 'assistant' | 'user';

export type ChatType = {
  role: ChatRole;
  content: string;
};

export interface ImageFile {
  base64Image: string;
  mimeType: string;
  size: number;
  name: string;
  width?: number;
  height?: number;
}

export interface FileData {
  base64Content: string;
  type: string;
  size: number;
  name: string;
}

export interface Message {
  question: string;
  answer: string;
  model: string;
}

export interface OpenAIChatContentImage {
  type: 'image_url';
  image_url: {
    url: string;
    detail: 'auto' | 'low' | 'high';
  };
}
export interface OpenAIChatContentText {
  type: 'text';
  text: string;
}
