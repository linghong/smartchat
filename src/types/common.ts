export interface OptionType {
  value: string;
  label: string;
  tags?: string[];
  category?: string;
  contextWindow?: number;
  vision?: boolean;
}

export interface ChatOption extends OptionType {
  tags: string[];
}

export type ApiResponse = {
  message: string;
  fileName: string;
  error?: string;
};

export type InputErrors = {
  [key: string]: string | null;
};

export type Input<T> = {
  [key: string]: T;
};

export type InputData = {
  [key: string]: string | number | null;
};

export type UploadData = {
  [key: string]: File | null;
};

export type UploadErrors = {
  [key: string]: string | null;
};
