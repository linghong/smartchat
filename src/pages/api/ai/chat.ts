import type { NextApiRequest, NextApiResponse } from 'next';
import { marked } from 'marked';

import { fetchDataFromPinecone } from '@/src/services/fetchDataFromPinecone';

import getClaudeChatCompletion from '@/src/services/llm/claude';
import getGeminiChatCompletion from '@/src/services/llm/gemini';
import { getGroqChatCompletion } from '@/src/services/llm/groq';
import {
  createEmbedding,
  getOpenAIChatCompletion
} from '@/src/services/llm/openai';
import getOpenModelChatCompletion from '@/src/services/llm/opensourceai';

import {
  processImageFiles,
  processNonMediaFiles
} from '@/src/utils/processMessageFile';
import { FileData, ImageFile } from '@/src/types/chat';
import {
  extractMessageContent,
  extractSubjectTitle
} from '@/src/utils/chatMessageHelper';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const { basePrompt, namespace, selectedModel, chatHistory, fileSrc } =
    req.body;

  let { question } = req.body;

  if (!question && fileSrc.length === 0) {
    return res.status(400).json({ message: 'No question in the request' });
  }

  if (!selectedModel?.value) {
    return res.status(500).json('Model name is missing.');
  }

  const base64ImageSrc = processImageFiles(fileSrc);

  question += await processNonMediaFiles(fileSrc);

  // Replace multiple newlines with a single newline to avoid excessive spacing
  const pretreatedQuestion = question.trim().replace(/\n\s*\n/g, '\n');

  try {
    let fetchedText = '';

    //fetch text from vector database
    if (namespace && namespace !== 'none') {
      const embeddedQuery = await createEmbedding(question);
      fetchedText = await fetchDataFromPinecone(embeddedQuery, namespace);
    }

    // get response from AI
    const { category } = selectedModel;
    let chatResponse: string | undefined | null;

    switch (category) {
      case 'openai':
        chatResponse = await getOpenAIChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'groq':
        chatResponse = await getGroqChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel
        );
        break;

      case 'google':
        chatResponse = await getGeminiChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'anthropic':
        chatResponse = await getClaudeChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'hf-small':
      case 'hf-large':
        const baseUrl =
          category === 'hf-small'
            ? process.env.NEXT_PUBLIC_SERVER_URL
            : process.env.NEXT_PUBLIC_SERVER_GPU_URL;
        if (!baseUrl) {
          return res
            .status(500)
            .json(
              `Url address for posting the data to ${selectedModel} is missing`
            );
        }
        const url = `${baseUrl}/api/chat_${category === 'hf-small' ? 'cpu' : 'gpu'}`;
        chatResponse = await getOpenModelChatCompletion(
          basePrompt,
          chatHistory,
          pretreatedQuestion,
          fetchedText,
          selectedModel,
          url
        );
        break;

      default:
        return res.status(500).json('Invalid model category');
    }

    let answer: string;
    let subject: string;
    if (!chatResponse) {
      answer = "Sorry, I'm having trouble finding an answer to your question.";
      subject = 'Unknown';
    } else {
      answer = extractMessageContent(chatResponse);
      subject = extractSubjectTitle(chatResponse);
    }

    const htmlAnswer = marked(answer);

    res.status(200).json({ answer: htmlAnswer, subject });
  } catch (error: any) {
    console.error('An error occurred: ', error);
    res.status(500).json({ error: error.message || 'Something went wrong' });
  }
}
