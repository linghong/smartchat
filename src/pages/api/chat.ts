import type { NextApiRequest, NextApiResponse } from 'next';
import { marked } from 'marked';

import { fetchDataFromPinecone } from '@/src/services/fetchDataFromPinecone';
import getGeminiChatCompletion from '@/src/services/llm/gemini';
import getClaudeChatCompletion from '@/src/services/llm/claude';
import { getGroqChatCompletion } from '@/src/services/llm/groq';
import {
  createEmbedding,
  getOpenAIChatCompletion
} from '@/src/services/llm/openai';
import getOpenModelChatCompletion from '@/src/services/llm/opensourceai';
import {
  extractMessageContent,
  extractSubjectTitle
} from '@/src/utils/chatMessageHelper';
import { ImageFile } from '@/src/types/chat';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const {
    basePrompt,
    question,
    namespace,
    selectedModel,
    chatHistory,
    imageSrc
  } = req.body;

  if (!question && imageSrc.length === 0) {
    return res.status(400).json({ message: 'No question in the request' });
  }

  if (!selectedModel?.value) {
    return res.status(500).json('Model name is missing.');
  }

  const base64ImageSrc = imageSrc?.map((imgSrc: ImageFile) => ({
    ...imgSrc,
    base64Image: imgSrc.base64Image.split(',')[1]
  }));

  // replacing newlines with spaces
  const sanitizedQuestion = question.trim().replaceAll('\n', ' ');

  try {
    let fetchedText = '';

    //fetch text from vector database
    if (namespace && namespace !== 'none') {
      const embeddedQuery = await createEmbedding(question);
      fetchedText = await fetchDataFromPinecone(embeddedQuery, namespace);
    }

    // get response from AI
    const { category } = selectedModel;
    let chatResponse: string | undefined | null;

    switch (category) {
      case 'openai':
        chatResponse = await getOpenAIChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'groq':
        chatResponse = await getGroqChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel
        );
        break;

      case 'google':
        chatResponse = await getGeminiChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'anthropic':
        chatResponse = await getClaudeChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          base64ImageSrc
        );
        break;

      case 'hf-small':
      case 'hf-large':
        const baseUrl =
          category === 'hf-small'
            ? process.env.NEXT_PUBLIC_SERVER_URL
            : process.env.NEXT_PUBLIC_SERVER_GPU_URL;
        if (!baseUrl) {
          return res
            .status(500)
            .json(
              `Url address for posting the data to ${selectedModel} is missing`
            );
        }
        const url = `${baseUrl}/api/chat_${category === 'hf-small' ? 'cpu' : 'gpu'}`;
        chatResponse = await getOpenModelChatCompletion(
          basePrompt,
          chatHistory,
          sanitizedQuestion,
          fetchedText,
          selectedModel,
          url
        );
        break;

      default:
        return res.status(500).json('Invalid model category');
    }

    let answer: string;
    let subject: string;
    if (!chatResponse) {
      answer = "Sorry, I'm having trouble finding an answer to your question.";
      subject = 'Unknown';
    } else {
      answer = extractMessageContent(chatResponse);
      subject = extractSubjectTitle(chatResponse);
    }

    const htmlAnswer = marked(answer);

    res.status(200).json({ answer: htmlAnswer, subject });
  } catch (error: any) {
    console.error('An error occurred: ', error);
    res.status(500).json({ error: error.message || 'Something went wrong' });
  }
}
