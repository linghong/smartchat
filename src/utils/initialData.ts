import { modelOptions } from '@/config/modellist';
import { AssistantOption } from '@/src/types/chat';

export const defaultModel = modelOptions[0];

export const defaultAssistants: AssistantOption[] = modelOptions.map(model => ({
  value: `default-${model.value}`,
  label: `Default ${model.label}`,
  isDefault: true,
  config: {
    name: `Default ${model.label}`,
    role: 'Default Assistant',
    model: model,
    topP: 0.7,
    temperature: 0.7,
    basePrompt: ''
  }
}));

export const initialMessage = {
  question: '',
  answer:
    'Hi! Please start by asking a question. You can click tag icons to add tags once the chat begins.',
  assistant: defaultAssistants[0].label
};
