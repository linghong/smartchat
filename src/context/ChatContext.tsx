'use client';
import React, {
  createContext,
  useContext,
  useState,
  ReactNode,
  Dispatch,
  SetStateAction
} from 'react';

import { Message, FileData } from '@/src/types/chat';
import { OptionType } from '@/src/types/common';
import { initialMessage, defaultModel } from '@/src/utils/initialData';

export interface ChatContextType {
  isNewChat: boolean;
  setIsNewChat: Dispatch<SetStateAction<boolean>>;
  isConfigPanelVisible: boolean;
  setIsConfigPanelVisible: Dispatch<SetStateAction<boolean>>;
  isSearchChat: boolean;
  setIsSearchChat: Dispatch<SetStateAction<boolean>>;
  fileSrcHistory: FileData[][];
  setFileSrcHistory: Dispatch<SetStateAction<FileData[][]>>;
  chatHistory: Message[];
  setChatHistory: Dispatch<SetStateAction<Message[]>>;
  activeChat: OptionType;
  setActiveChat: Dispatch<SetStateAction<OptionType>>;
  chats: OptionType[];
  setChats: Dispatch<SetStateAction<OptionType[]>>;
  selectedModel: OptionType;
  setSelectedModel: Dispatch<SetStateAction<OptionType>>;
}

export const ChatContext = createContext<ChatContextType | undefined>(
  undefined
);

export const ChatProvider: React.FC<{ children: ReactNode }> = ({
  children
}) => {
  const [isNewChat, setIsNewChat] = useState(true);
  const [isConfigPanelVisible, setIsConfigPanelVisible] = useState(false);
  const [isSearchChat, setIsSearchChat] = useState(false);
  const [fileSrcHistory, setFileSrcHistory] = useState<FileData[][]>([[]]);
  const [chatHistory, setChatHistory] = useState<Message[]>([initialMessage]);
  const [activeChat, setActiveChat] = useState<OptionType>({
    value: '0',
    label: '0',
    tags: []
  }); //new chat
  const [chats, setChats] = useState<OptionType[]>([]);
  const [selectedModel, setSelectedModel] = useState<OptionType>(defaultModel);

  return (
    <ChatContext.Provider
      value={{
        isNewChat,
        setIsNewChat,
        isConfigPanelVisible,
        setIsConfigPanelVisible,
        isSearchChat,
        setIsSearchChat,
        fileSrcHistory,
        setFileSrcHistory,
        chatHistory,
        setChatHistory,
        activeChat,
        setActiveChat,
        chats,
        setChats,
        selectedModel,
        setSelectedModel
      }}
    >
      {children}
    </ChatContext.Provider>
  );
};

export const useChatContext = () => {
  const context = useContext(ChatContext);
  if (context === undefined) {
    throw new Error('useChatContext must be used within a ChatProvider');
  }
  return context;
};
