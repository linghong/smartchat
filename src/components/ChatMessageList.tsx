import React, { useRef, useEffect } from 'react';

import ChatMessage from '@/src/components/ChatMessage';
import { useChatContext } from '@/src/context/ChatContext';
import { AssistantOption } from '@/src/types/chat';
import { initialMessage } from '@/src/utils/initialData';

export interface ChatMessageListProps {
  loading: boolean;
  assistantOptions: AssistantOption[];
  selectedAssistant: AssistantOption;
  handleAssistantChange: (newValue: AssistantOption) => void;
  handleFileDelete: (id: number) => void;
  handleRetry?: () => void;
  handleCopy: () => void;
}

const ChatMessageList: React.FC<ChatMessageListProps> = ({
  assistantOptions,
  selectedAssistant,
  handleAssistantChange,
  loading,
  handleFileDelete,
  handleRetry,
  handleCopy
}) => {
  const messagesRef = useRef<HTMLDivElement>(null);

  const { fileSrcHistory, chatHistory } = useChatContext();

  useEffect(() => {
    if (messagesRef.current) {
      messagesRef.current.scrollTop = messagesRef.current.scrollHeight;
    }
  }, [chatHistory]);

  const isNew = chatHistory[0].answer === initialMessage.answer;

  return (
    <div className="flex-grow overflow-y-auto bg-white border-2 border-stone-200">
      <div
        className="w-full h-full rounded-lg"
        aria-live="polite"
        aria-atomic="true"
        ref={messagesRef}
      >
        {chatHistory.map((chat, index) => (
          <ChatMessage
            key={index}
            isNew={isNew && index === 0}
            message={chat}
            lastIndex={index === chatHistory.length - 1}
            loading={loading}
            fileSrc={fileSrcHistory[index]}
            selectedAssistant={selectedAssistant}
            handleAssistantChange={handleAssistantChange}
            assistantOptions={assistantOptions}
            handleFileDelete={handleFileDelete}
            handleCopy={handleCopy}
            handleRetry={
              index === chatHistory.length - 1 ? handleRetry : undefined
            }
          />
        ))}
      </div>
    </div>
  );
};

export default ChatMessageList;
