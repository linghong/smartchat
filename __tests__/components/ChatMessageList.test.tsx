import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import ChatMessageList, {
  ChatMessageListProps
} from '@/src/components/ChatMessageList';
import { FileData, AssistantOption } from '@/src/types/chat';
import { defaultAssistants } from '@/src/utils/initialData';
import { renderWithContext } from '@/__tests__/test_utils/context';

jest.mock('@/src/components/ChatMessage', () => {
  return function MockChatMessage({
    message,
    isNew,
    lastIndex,
    loading,
    fileSrc,
    selectedAssistant,
    handleAssistantChange,
    assistantOptions,
    handleCopy,
    handleRetry,
    handleFileDelete
  }: any) {
    return (
      <div data-testid={`chat-message`}>
        {!isNew && <div className="user">{message.question}</div>}
        <div className="ai">
          <div>{message.model}</div>
          {lastIndex && loading ? (
            <div aria-label="loading">Just a moment, Iâ€™m working on it...</div>
          ) : (
            <div>{message.answer}</div>
          )}
          {fileSrc &&
            fileSrc.map((file: FileData, i: number) => (
              <div key={i}>
                {file.name} ({file.type})
                <button onClick={() => handleFileDelete(i)}>Delete</button>
              </div>
            ))}
          {!isNew && lastIndex && (
            <div>
              {handleRetry && <button onClick={handleRetry}>Retry</button>}
              {handleCopy && message.answer !== '' && (
                <button onClick={handleCopy}>Copy</button>
              )}
              <select>
                {assistantOptions.map((assistant: AssistantOption) => (
                  <option
                    key={assistant.value}
                    onClick={() => handleAssistantChange(assistant)}
                  >
                    {assistant.label}
                  </option>
                ))}
              </select>
            </div>
          )}
        </div>
      </div>
    );
  };
});

describe('ChatMessageList', () => {
  const mockHandleFileDelete = jest.fn();
  const mockHandleAssistantChange = jest.fn();
  const mockHandleRetry = jest.fn();
  const mockHandleCopy = jest.fn();

  const defaultProps: ChatMessageListProps = {
    loading: false,
    selectedAssistant: defaultAssistants[0],
    assistantOptions: defaultAssistants,
    handleAssistantChange: mockHandleAssistantChange,
    handleFileDelete: mockHandleFileDelete,
    handleRetry: mockHandleRetry,
    handleCopy: mockHandleCopy
  };

  beforeEach(() => {
    jest.clearAllMocks();
    localStorage.clear();
  });

  it('renders correct number of ChatMessage components', () => {
    renderWithContext(<ChatMessageList {...defaultProps} />);
    const chatMessages = screen.getAllByTestId('chat-message');
    expect(chatMessages).toHaveLength(2);
  });

  it('renders correctly with chat history', () => {
    renderWithContext(<ChatMessageList {...defaultProps} />);
    expect(screen.getByText('How are you?')).toBeInTheDocument();
    expect(screen.getByText("I'm doing well, thank you!")).toBeInTheDocument();
  });

  it('renders correctly when loading', () => {
    renderWithContext(<ChatMessageList {...defaultProps} loading={true} />);
    expect(screen.getByLabelText('loading')).toBeInTheDocument();
  });

  it('handles file src history and delete', () => {
    const fileSrcHistory = [
      [
        {
          base64Content: 'data:image/png;base64,base64Imagetest1',
          type: 'image/png',
          size: 2 * 1024 * 1024, //2MB
          name: 'test1.png'
        }
      ],
      [
        {
          base64Content: 'data:image/png;base64,base64Imagetest2',
          type: 'image/jpeg',
          size: 2 * 1024 * 1024, //2MB
          name: 'test2.jpg'
        }
      ]
    ];

    renderWithContext(<ChatMessageList {...defaultProps} />, {
      fileSrcHistory
    });

    expect(screen.getByText('test1.png (image/png)')).toBeInTheDocument();
    expect(screen.getByText('test2.jpg (image/jpeg)')).toBeInTheDocument();

    const deleteButtons = screen.getAllByText('Delete');
    expect(deleteButtons).toHaveLength(2);
    fireEvent.click(deleteButtons[0]);
    expect(mockHandleFileDelete).toHaveBeenCalledWith(0);
  });

  it('indicates a new conversation', () => {
    const newChatMessageListContext = {
      chatHistory: [
        {
          question: '',
          answer: 'Hi, how can I assist you?',
          assistant: 'GPT-4 Assistant'
        }
      ],
      firSrcHistory: [[]]
    };
    renderWithContext(<ChatMessageList {...defaultProps} />, {
      ...newChatMessageListContext
    });

    expect(screen.getByText('Hi, how can I assist you?')).toBeInTheDocument();
    expect(screen.queryByTestId('chat-message')).not.toHaveTextContent('');
  });

  it('renders retry button for the last message', () => {
    renderWithContext(<ChatMessageList {...defaultProps} />);
    const retryButton = screen.getByText('Retry');
    expect(retryButton).toBeInTheDocument();
  });

  it('calls handleRetry when Retry button is clicked', () => {
    renderWithContext(<ChatMessageList {...defaultProps} />);
    const retryButton = screen.getByText('Retry');
    fireEvent.click(retryButton);
    expect(mockHandleRetry).toHaveBeenCalled();
  });

  it('renders copy button for the last message', () => {
    renderWithContext(<ChatMessageList {...defaultProps} />);
    const copyButton = screen.getByText('Copy');
    expect(copyButton).toBeInTheDocument();
  });

  it('calls handleCopy when Copy button is clicked', () => {
    renderWithContext(<ChatMessageList {...defaultProps} />);
    const copyButton = screen.getByText('Copy');
    fireEvent.click(copyButton);
    expect(mockHandleCopy).toHaveBeenCalled();
  });

  it('handles assistant change', () => {
    renderWithContext(<ChatMessageList {...defaultProps} />);
    const combobox = screen.getByRole('combobox');
    expect(combobox).toBeInTheDocument();
    expect(combobox).toHaveTextContent('Default Gemini-1.5 Pro Exp');

    // simulate the selection process
    fireEvent.click(combobox);
    const options = screen.getAllByRole('option');
    fireEvent.click(options[1]); // Select the second option

    expect(mockHandleAssistantChange).toHaveBeenCalledWith(
      defaultAssistants[1]
    );
  });

  it('matches snapshot', () => {
    const { asFragment } = renderWithContext(
      <ChatMessageList {...defaultProps} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
