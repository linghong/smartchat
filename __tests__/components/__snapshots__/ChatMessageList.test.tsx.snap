// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessageList matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex-grow overflow-y-auto bg-white border-2 border-stone-200"
    style="height: calc(100vh - 200px); scroll-behavior: smooth;"
  >
    <div
      aria-atomic="true"
      aria-live="polite"
      class="w-full rounded-lg"
    >
      <div
        data-testid="chat-message"
      >
        <div
          class="user"
        >
          Hello
        </div>
        <div
          class="ai"
        >
          <div />
          <div>
            How can I assistant you?
          </div>
        </div>
      </div>
      <div
        data-testid="chat-message"
      >
        <div
          class="user"
        >
          How are you?
        </div>
        <div
          class="ai"
        >
          <div />
          <div>
            I'm doing well, thank you!
          </div>
          <div>
            <button>
              Retry
            </button>
            <button>
              Copy
            </button>
            <select>
              <option>
                Default Gemini-1.5 Pro Exp
              </option>
              <option>
                Default Gemini-1.5 Flash Exp
              </option>
              <option>
                Default Gemini-1.5 Flash-8b Exp
              </option>
              <option>
                Default Gemini-1.5 Pro
              </option>
              <option>
                Default Gemini-1.5 Flash
              </option>
              <option>
                Default Claude 3.5 Sonnet
              </option>
              <option>
                Default Claude 3 Haiku
              </option>
              <option>
                Default Claude 3 Opus
              </option>
              <option>
                Default GPT-4o-2024-08-06
              </option>
              <option>
                Default GPT-4o
              </option>
              <option>
                Default GPT-4o Mini
              </option>
              <option>
                Default GPT-4 Turbo
              </option>
              <option>
                Default GPT-4
              </option>
              <option>
                Default Llama3.2 90b
              </option>
              <option>
                Default Llama3.2 3b
              </option>
              <option>
                Default Llama3.2 1b
              </option>
              <option>
                Default Llama3.1 70b
              </option>
              <option>
                Default Llama3.1 8b
              </option>
              <option>
                Default Llama3 70b
              </option>
              <option>
                Default Llama3 8b
              </option>
              <option>
                Default Mixtral 8x7b
              </option>
              <option>
                Default Gemma 7b
              </option>
              <option>
                Default Gemma2 9b
              </option>
              <option>
                Default Llama-2-7b-chat-hf
              </option>
              <option>
                Default phi-1_5
              </option>
            </select>
          </div>
        </div>
      </div>
      <div />
    </div>
  </div>
</DocumentFragment>
`;
